examples/CustomCollectionView-Swift/Sample/MosaicCollectionViewLayout.swift

```diff
@@ -0,0 +1,254 @@
+//
+//  MosaicCollectionViewLayout
+//  Sample
+//
+//  Created by Rajeev Gupta on 11/9/16.
+//
+//  Copyright (c) 2014-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under the BSD-style license found in the
+//  LICENSE file in the root directory of this source tree. An additional grant
+//  of patent rights can be found in the PATENTS file in the same directory.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+//  FACEBOOK BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+//  ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+//  CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+import Foundation
+import UIKit
+import AsyncDisplayKit
+
+protocol MosaicCollectionViewLayoutDelegate: ASCollectionDelegate {
+  func collectionView(_ collectionView: UICollectionView, layout: MosaicCollectionViewLayout, originalItemSizeAtIndexPath: IndexPath) -> CGSize
+}
+
+class MosaicCollectionViewLayout: UICollectionViewFlowLayout {
+  var numberOfColumns: Int
+  var columnSpacing: CGFloat
+  var _sectionInset: UIEdgeInsets
+  var interItemSpacing: UIEdgeInsets
+  var headerHeight: CGFloat
+  var _columnHeights: [[CGFloat]]?
+  var _itemAttributes: [[UICollectionViewLayoutAttributes]]?
+  var _headerAttributes: [UICollectionViewLayoutAttributes]?
+  var _allAttributes: [UICollectionViewLayoutAttributes]?
```

Comment:
Those attributes could be initialized as empty arrays, example:
`var _allAttributes = [UICollectionViewLayoutAttributes]()`
This way you don't need to do any optional unwrapping later on. I didn't see a reason why these would/should be nil